src/Interpreters/TraceCollector.cpp

```diff
@@ -34,32 +34,37 @@ TraceCollector::~TraceCollector()
 {
     try
     {
-        if (!thread.joinable())
-            LOG_ERROR(&Poco::Logger::get("TraceCollector"), "TraceCollector thread is malformed and cannot be joined");
+        if (thread.joinable())
+        {
+            /** Sends TraceCollector stop message
+            *
+            * Each sequence of data for TraceCollector thread starts with a boolean flag.
+            * If this flag is true, TraceCollector must stop reading trace_pipe and exit.
+            * This function sends flag with a true value to stop TraceCollector gracefully.
+            */
+            WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
+            writeChar(true, out);
+            out.next();
+        }
         else
-            stop();
+            LOG_ERROR(&Poco::Logger::get("TraceCollector"), "TraceCollector thread is malformed and cannot be joined");
+    }
+    catch (...)
+    {
+        tryLogCurrentException("TraceCollector");
+    }
 
+    try
+    {
         TraceSender::pipe.close();
```

Comment:
Better, the only problem I see is if `close` fails but I have no idea what we can do here if there is no way to forcefully break the pipe no matter the status.

Reply:
yes, this problem pretty much a vicious cycle... and I would like to see your thoughts on possible exception in `TraceCollector::run()` loop - the problem here is that for any exception this thread will be ended, and if it's not due to broken pipe, producer will eventually fill up pipe buffer and produce bunch of exceptions with CannotWriteToWriteBufferDiscard telemetry - which can be seen as annoying and maybe unnecessary... but frankly I have no idea what is better solution here...